utils/pattern/src/parser/mod.rs

```diff
@@ -0,0 +1,315 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+pub mod error;
+
+use crate::token::PatternToken;
+pub use error::ParserError;
+use std::{fmt::Debug, str::FromStr};
+
+#[derive(PartialEq)]
+enum ParserState {
+    Default,
+    Placeholder,
+    QuotedLiteral,
+}
+
+impl Default for ParserState {
+    fn default() -> Self {
+        Self::Default
+    }
+}
+
+macro_rules! advance_state {
+    ($self:ident, $idx:expr, $next_state:path) => {{
+        let range = $self.start_idx..$idx;
+        $self.idx = $idx + 1;
+        $self.start_idx = $self.idx;
+        $self.state = $next_state;
+        range
+    }};
+}
+
+macro_rules! handle_literal {
+    ($self:ident, $quoted:expr, $next_state:path) => {{
+        let range = advance_state!($self, $self.idx, $next_state);
+        if !range.is_empty() {
+            return Ok(Some(PatternToken::Literal {
+                content: &$self.input[range],
+                quoted: $quoted,
+            }));
+        } else {
+            continue;
```

Comment:
Issue: If the range is empty, you could be in a double-quote quoted literal, i.e. `''`.  This resolves to a single quote.  For example, the patterns `o''clock` and `'o''clock'` both map to the string `o'clock`.  Please add tests for two quotes in a row.

Reply:
Great catch! Thank you!
I haven't backported https://github.com/zbraniecki/icu4x/blob/pattern_formatter2/components/datetime/src/pattern/parser.rs#L81 when extracting this util. I'll add it!
And yes, we have tests for `''` which I'll add here.

Reply:
Done (will add many more tests once we agree on the general shape of the API)